### PR TITLE
Change `highlightRange` to accept an ordinary DOM Range

### PR DESCRIPTION
**Pull request by robertknight**
This PR is preparation for some upcoming changes to fix bugs related to overlapping highlights (possibly the original cause of https://github.com/hypothesis/product-backlog/issues/954, but see also https://github.com/hypothesis/client/issues/2326, https://github.com/hypothesis/h/issues/5997). The issue in these cases relates to the fact that the client first anchors annotations to ranges and then adds highlights. The process of inserting the highlights for an anchor modifies the DOM and that can alter the ranges of other anchors. My plan to resolve this is to change the representation of anchored ranges to one which is robust to the kinds of DOM modification made by highlighting [1]. This new representation of an anchor's range will be _resolved_ to a concrete DOM Range just before the highlights are inserted.

As a preliminary step, this PR changes the `highlightRange` function which wraps the contents of an anchor's range with highlights to work with an ordinary DOM `Range` instead of the `NormalizedRange` wrapper class inherited from annotator. In the process I found a bug in one of the helper functions in `range-util.js`, `isNodeInRange`, and rewrote the implementation and added some extra tests to fix that.

----

[1] Specifically, my plan is to represent anchored ranges as a pair of (element, textPosition) points representing the start and end of the range, where `element` refers to the first non-highlight parent element of the range's start/end point and `textPosition` is the character offset in `element`'s text content. If the DOM structure is changed in between anchoring and inserting highlights for a particular annotation, it will still be possible to resolve this representation to the same content, as long as the text content hasn't changed.

### CODE REVIEW

File: src/annotator/test/range-util-test.js [hunk 1]
```diff
@@ -44,26 +44,39 @@ describe('annotator.range-util', function () {
...
+      range.setStart(node, 0);
+      range.setEnd(node, 0);
+      assert.isTrue(rangeUtil.isNodeInRange(range, node));
+    });
+
+    it('can test a node against an empty range', () => {
+      const node = document.createElement('span');
+      const range = new Range();
+      assert.isFalse(rangeUtil.isNodeInRange(range, node));
+    });
```
Comment (robertknight) [inline src/annotator/test/range-util-test.js:79]: Both of these cases used to result in an exception being thrown.

File: src/annotator/range-util.js [hunk 2]
```diff
@@ -14,49 +14,43 @@ export function isSelectionBackwards(selection) {
...
+      // Check start of node is before end of range.
+      range.comparePoint(node, 0) <= 0 &&
+      // Check end of node is after start of range.
+      range.comparePoint(node, length) >= 0
+    );
+  } catch (e) {
+    // `comparePoint` may fail if the `range` and `node` do not share a common
+    // ancestor or `node` is a doctype.
+    return false;
   }
```
Comment (robertknight) [inline src/annotator/range-util.js:35]: [`Range.comparePoint`](https://developer.mozilla.org/en-US/docs/Web/API/Range/comparePoint) is available in Edge 17 and above, as well as all other engines. This fortunately matches our new browser support baseline :)

File: src/annotator/highlighter.js [hunk 3]
```diff
@@ -150,19 +144,48 @@ function drawHighlightsAbovePdfCanvas(highlightEl) {
  * @typedef {HTMLElement & HighlightProps} HighlightElement
  */
 
+/**
+ * Return the complete text nodes inside `range`.
```
Comment (lyzadanger) [inline src/annotator/highlighter.js:148]: "complete text nodes"... did you mean "complete set of text nodes" here or "the set of all complete text nodes"?

File: src/annotator/highlighter.js [hunk 4]
```diff
@@ -150,19 +144,48 @@ function drawHighlightsAbovePdfCanvas(highlightEl) {
...
+ *
+ * If a range starts or ends part-way through a text node, the nodes are split
```
Comment (lyzadanger) [inline src/annotator/highlighter.js:150]: "the nodes"...meaning the text node that is only partially covered by the range?

File: src/annotator/highlighter.js [hunk 5]
```diff
@@ -150,19 +144,48 @@ function drawHighlightsAbovePdfCanvas(highlightEl) {
...
+ * @return {Text[]}
+ */
+function wholeTextNodesInRange(range) {
+  const textNodes = [];
+
+  forEachNodeInRange(range, node => {
+    if (node.nodeType !== Node.TEXT_NODE) {
+      return;
+    }
+    let text = /** @type {Text} */ (node);
```
Comment (lyzadanger) [inline src/annotator/highlighter.js:163]: We chatted a little in Slack: I think it would be helpful to drop a comment or two in this logic. You're using legit web APIs here, but I did have to spend a few minutes on MDN's docs about [`splitText`](https://developer.mozilla.org/en-US/docs/Web/API/Text/splitText) to understand what was going on. i.e. I know any comments here would be technically superfluous but might save someone a little time later, especially if they captured the _intent_ of the logic.

File: src/annotator/highlighter.js [hunk 6]
```diff
@@ -150,19 +144,48 @@ function drawHighlightsAbovePdfCanvas(highlightEl) {
...
+
+  return textNodes;
+}
+
 /**
  * Wraps the DOM Nodes within the provided range with a highlight
  * element of the specified class and returns the highlight Elements.
  *
- * @param {NormalizedRange} normedRange - Range to be highlighted.
+ * @param {Range} range - Range to be highlighted.
```
Comment (lyzadanger) [inline src/annotator/highlighter.js:183]: Nit: you might consider doing away with the full stop here (I know you inherited it)

File: src/annotator/highlighter.js [hunk 7]
```diff
@@ -191,6 +214,7 @@ export function highlightRange(normedRange, cssClass = 'hypothesis-highlight') {
   // Filter out text node spans that consist only of white space. This avoids
   // inserting highlight elements in places that can only contain a restricted
   // subset of nodes such as table rows and lists.
+  const white = /^\s*$/;
```
Comment (lyzadanger) [inline src/annotator/highlighter.js:217]: I realize this is inherited, but `whitespace` would be a nicer name here.

File: src/annotator/range-util.js [hunk 8]
```diff
@@ -14,49 +14,43 @@ export function isSelectionBackwards(selection) {
...
- * with IE.
+ * Returns true if any part of `node` lies within `range`.
  *
  * @param {Range} range
  * @param {Node} node
  */
 export function isNodeInRange(range, node) {
-  if (node === range.startContainer || node === range.endContainer) {
-    return true;
+  try {
```
Comment (lyzadanger) [inline src/annotator/range-util.js:23]: I note that MDN still marks `comparePoint` as "experimental." It's supported in the browsers we care about, but I thought I'd ask after that...

Edited: I see _your_ comment about support in Edge 17+ ... my question is broader now: what's our stance on "experimental" APIs, if any?

File: src/annotator/range-util.js [hunk 9]
```diff
@@ -14,49 +14,43 @@ export function isSelectionBackwards(selection) {
...
+    const length = node.nodeValue?.length ?? node.childNodes.length;
+    return (
+      // Check start of node is before end of range.
+      range.comparePoint(node, 0) <= 0 &&
+      // Check end of node is after start of range.
+      range.comparePoint(node, length) >= 0
+    );
+  } catch (e) {
+    // `comparePoint` may fail if the `range` and `node` do not share a common
```
Comment (lyzadanger) [inline src/annotator/range-util.js:32]: Thanks for this comment. I was already wondering...

File: src/annotator/range-util.js (hunk 2 above)
Comment (lyzadanger) [reply src/annotator/range-util.js:35]: Oh, you answered my question!

File: src/annotator/range-util.js [hunk 10]
```diff
@@ -14,49 +14,43 @@ export function isSelectionBackwards(selection) {
...
-  const isAtOrAfterEnd =
-    range.compareBoundaryPoints(Range.END_TO_END, nodeRange) >= 0;
-  nodeRange.detach();
-  return isAtOrBeforeStart && isAtOrAfterEnd;
 }
 
 /**
- * Iterate over all Node(s) in `range` in document order and invoke `callback`
- * for each of them.
+ * Iterate over all Node(s) which overlap `range` in document order and invoke
```
Comment (lyzadanger) [inline src/annotator/range-util.js:39]: "Overlap" here meaning "are at least partially within?" Or something else?

File: src/annotator/test/highlighter-test.js [hunk 11]
```diff
@@ -54,12 +52,10 @@ function PdfPage({ showPlaceholder = false }) {
...
-  const r = new NormalizedRange({
-    commonAncestor: textSpan,
-    start: textSpan.childNodes[0],
-    end: textSpan.childNodes[0],
-  });
-  return highlightRange(r);
+  const range = new Range();
+  range.setStartBefore(textSpan.childNodes[0]);
+  range.setEndAfter(textSpan.childNodes[0]);
+  return highlightRange(range);
```
Comment (lyzadanger) [inline src/annotator/test/highlighter-test.js:58]: Thanks for `range` instead of `r` here.

File: src/annotator/test/highlighter-test.js [hunk 12]
```diff
@@ -79,16 +75,14 @@ function createPdfPageWithHighlight() {
 describe('annotator/highlighter', () => {
   describe('highlightRange', () => {
     it('wraps a highlight span around the given range', () => {
-      const txt = document.createTextNode('test highlight span');
+      const text = document.createTextNode('test highlight span');
```
Comment (lyzadanger) [inline src/annotator/test/highlighter-test.js:78]: Thanks for adding the `e` (seriously)

File: src/annotator/highlighter.js [hunk 13]
```diff
@@ -150,19 +144,48 @@ function drawHighlightsAbovePdfCanvas(highlightEl) {
...
+  const textNodes = [];
+
+  forEachNodeInRange(range, node => {
+    if (node.nodeType !== Node.TEXT_NODE) {
+      return;
+    }
+    let text = /** @type {Text} */ (node);
+
+    if (text === range.startContainer && range.startOffset > 0) {
+      text = text.splitText(range.startOffset);
```
Comment (lyzadanger) [inline src/annotator/highlighter.js:166]: This line appears to be a test miss. Can you rectify that?

File: src/annotator/test/range-util-test.js (hunk 1 above)
Comment (lyzadanger) [reply src/annotator/test/range-util-test.js:79]: Uncaught?

File: src/annotator/test/range-util-test.js [hunk 14]
```diff
@@ -44,26 +44,39 @@ describe('annotator.range-util', function () {
     selection.addRange(range);
   }
 
-  describe('#isNodeInRange', function () {
-    it('is true for a node in the range', function () {
-      const rng = createRange(testNode, 0, 1);
```
Comment (lyzadanger) [inline src/annotator/test/range-util-test.js:49]: Seriously, thank you.

**Review by lyzadanger — DISMISSED** [review]
This looks good as-is. I only requested changes because the test coverage feels light. I realize much of that coverage laxity is inherited, so some of it might be busywork/ignore-able, but I did flag a miss in newly-added code that might be good to cover. Feel free to merge once you're happy with test coverage.

I also flagged a few confusing spots that might be aided by tweaking/adding comments.

File: src/annotator/highlighter.js (hunk 13 above)
Comment (robertknight) [reply src/annotator/highlighter.js:166]: Indeed! - I've added a bunch of extra tests and caught a bug in the process.

File: src/annotator/highlighter.js (hunk 5 above)
Comment (robertknight) [reply src/annotator/highlighter.js:163]: Done. I've added comments to clarify.

File: src/annotator/highlighter.js (hunk 4 above)
Comment (robertknight) [reply src/annotator/highlighter.js:150]: This should have said "the node is split". I'll correct it.

File: src/annotator/highlighter.js (hunk 3 above)
Comment (robertknight) [reply src/annotator/highlighter.js]: Ah. I meant "Return all the text nodes which are wholly inside range". I've rewritten this comment.

File: src/annotator/highlighter.js (hunk 6 above)
Comment (robertknight) [reply src/annotator/highlighter.js]: Done

File: src/annotator/highlighter.js (hunk 7 above)
Comment (robertknight) [reply src/annotator/highlighter.js]: Done

File: src/annotator/range-util.js (hunk 8 above)
Comment (robertknight) [reply src/annotator/range-util.js:23]: The information on MDN is out of date as far as I can tell. The compatibility information on [caniuse](https://caniuse.com/mdn-api_range_comparepoint) is incomplete, but this appears to be because the API is so old. I found references to it in WebKit in 2008 and a Firefox bug report about some spec compliance details in 2013. An API that has existed in multiple major browsers for that long will never get breaking changes unless it causes a major security or performance problem and nobody is using it.

Our general policy has been that once an API (or ES language feature) ships on-by-default in the stable version of two major browsers it is safe to use. Special consideration might have to be given to very complex features (eg. Service Workers or Web Components).

File: src/annotator/range-util.js (hunk 10 above)
Comment (robertknight) [reply src/annotator/range-util.js:39]: Yes, "at least partially within".
